specs/experimental/gov-delegation.md

```diff
@@ -0,0 +1,449 @@
+
+### Differences
+
+The main difference for delegations made from the `GovernanceToken` contract is that basic delegations are encapsulated in
+a [`Delegation`](#delegation) struct and forwarded to the `GovernanceDelegation` contract. Basic delegation can be achieved
+with a `Delegation` such as:
+
+```solidity
+Delegation({
+  delegatee: ...,
+  allowanceType: AllowanceType.Relative,
+  allowance: 10e4 // 100%
+})
+```
+
+The following diagram shows the sequence of a basic delegation performed from the `GovernanceToken` contract.
+
+```mermaid
+sequenceDiagram
+    User ->> GovernanceToken: call `delegate` or `delegateBySig`
+    GovernanceToken ->> GovernanceDelegation: call `delegateFromToken`
+    GovernanceDelegation -->> GovernanceToken: migrate if user hasn't been migrated
+    GovernanceDelegation ->> GovernanceDelegation: apply basic delegation
+```
+
+Once a user has been migrated to the `GovernanceDelegation`, the `GovernanceToken` MUST always use the `GovernanceDelegation`'s
+delegation state. The following diagram shows the control flow for this case.
+
+```mermaid
+flowchart TD
+    A1(User) -->|Calls delegation getter function | A2(GovernanceToken)
+    A2 --> A3{Is user migrated?}
+    A3 --> |Yes, fetch from GovernanceDelegation| A4(GovernanceDelegation)
+    A3 --> |No, use its state| A2
+    A2 --> |Return data| A1
+```
+
+## Delegation Validation
+
+When applying a new delegation set as part of the [`_delegate`](#_delegate) function, the `GovernanceDelegation` MUST check
+that:
+
+- The length of the `_newDelegations` array DOES NOT exceed `MAX_DELEGATIONS`.
+- The sum of `amount` in relative delegations DOES NOT exceed the `DENOMINATOR`, and each `amount` is greater than 0.
+- The sum of `amount` in absolute delegations DOES NOT exceed the total voting power of the `_delegator`.
+- The new delegatee set is sorted in descending order by voting power.
+
+If any of the above conditions are not met, the `_delegate` function MUST revert with a specific error message.
+
+## Security Considerations
+
+### Dependence on GovernanceDelegation
+
+As the `GovernanceToken` depends on the `GovernanceDelegation` contract, the `GovernanceDelegation` contract MUST be implemented
+so that it minimizes the risk of unexpected reverts during the transfer hook call. If the `GovernanceDelegation` contract
+reverts, `GovernanceToken` transfers will be blocked. Additionally, the `GovernanceToken` MUST always use the
+`GovernanceDelegation`'s delegation state if a user has been migrated.
+
+### Connection with GovernanceToken
+
+Similarly, the `GovernanceDelegation` MUST always process token transfers from the `GovernanceToken` contract correctly to
+stay in sync with token balances. If the `GovernanceDelegation` contract is not in sync with the `GovernanceToken` contract,
+the voting power of users MAY be incorrect or outdated.
+
+## Future Considerations
```

Comment:
I think we can drop this section, we'll be working on that spec right after this one anyway.